### PR TITLE
Multi instance Firmware Object support

### PR DESCRIPTION
**Pull request by jheiskan81**
### Full modem config init update

dfu_target_full_modem_cfg() now copy parametrs and those are used by init stream at dfu_target_full_modem_init().

### Fota download client update

**Config update: FOTA_CLIENT_EVT_DONE_SHEDULE_DFU**

Added possible to skip automated firmware schedule permission
at finished push or pull operation. Default is behavior is original.

Bug Fix: Fix crash if fota_cancel is called until any start.

### Advanced Firmware Object
New object Support for multi instace object.
* Enable possibility ctreate Own Instance for Modem and Application.
 
New Cancel Execution support for cancel update process.

New resources:
* Last state change Update timestamp
* Current image version
* FOTA Component name
* Image current version
* Maxium Defer period for move update process

### LwM2M Firmware update utils

Advanced firmware object multi instance update
* Wrapper layer for old FOTA and new advanced FOTA
* create own Instance for Modem and Application
 
Update execution update:
* Init DFU by Image type
* mark DFU firmware shedule to target.

**Reset and Cancel execution update**
* Init DFU by Image type
* Cancel operation is accept also from downloading state.

Storing Firmware objects critical parameters to settings
* Firmware object State
* Firmware update result
* DFU image type
* Remove old settings usage for counter.

### LwM2M sample update

**Overlay for advanced firmware**
* New firmware for use LwM2M multi instance Fota object

Removed old Fota settings init and include file

### Asset Tracker V2

Removed Extra callcack for Update state and use State callback for
trigger CLOUD_WRAP_EVT_FOTA_DONE event.

Rework logging in modules thereby reducing flash usage.

### CODE REVIEW

File: samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CONFIG_LWM2M_CLIENT_UTILS=y
```
Comment (VeijoPesonen) [inline samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf:1]: This new overlay should be described on the sample documentation.

File: include/net/fota_download.h [hunk 2]
```diff
@@ -83,6 +83,14 @@ struct fota_download_evt {
  */
 typedef void (*fota_download_callback_t)(const struct fota_download_evt *evt);
 
+/**
+ * @brief FOTA download pre dfu taget init callback function.
```
Comment (VeijoPesonen) [inline include/net/fota_download.h:87]:
```suggestion
 * @brief FOTA download pre dfu target init callback function.
```

File: include/net/fota_download.h [hunk 3]
```diff
@@ -92,6 +100,17 @@ typedef void (*fota_download_callback_t)(const struct fota_download_evt *evt);
  */
 int fota_download_init(fota_download_callback_t client_callback);
 
+/**@brief Set the firmware over-the-air download library dfu targetn init callback.
```
Comment (VeijoPesonen) [inline include/net/fota_download.h:103]:
```suggestion
/**@brief Set the firmware over-the-air download library dfu target init callback.
```

File: include/net/fota_download.h [hunk 4]
```diff
@@ -92,6 +100,17 @@ typedef void (*fota_download_callback_t)(const struct fota_download_evt *evt);
...
+ *
+ * Register callback client user can prepare dfu target device ready for init.
+ *
+ * @param cb Callback for called before dfu_target_init
```
Comment (VeijoPesonen) [inline include/net/fota_download.h:107]:
```suggestion
 * @param cb The callback.
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+
+/* resource state variables */
+static uint8_t update_state[MAX_INSTANCE_COUNT];
+static uint8_t update_result[MAX_INSTANCE_COUNT];
+static uint8_t delivery_method[MAX_INSTANCE_COUNT];
+static time_t last_change[MAX_INSTANCE_COUNT];
+static char package_uri[MAX_INSTANCE_COUNT][PACKAGE_URI_LEN];
+static char component_version[MAX_INSTANCE_COUNT][PACKAGE_URI_LEN];
+
+/* A varying number of firmware object exists */
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c:69]:
```suggestion
/* A varying number of firmware object instances exists */
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	uint8_t state;
+	int ret = 0;
+	lwm2m_engine_set_data_cb_t callback;
+
+	state = lwm2m_adv_firmware_get_update_state(obj_inst_id);
+	if (state == STATE_IDLE) {
+		/* TODO: setup timer to check download status,
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c:249]: I assume these TODOs are remnants from the object 5.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 7]
```diff
@@ -69,16 +94,285 @@ static void on_modem_lib_init(int ret, void *ctx)
...
+	ret = snprintk(path, sizeof(path), "%d/%d/%d", ENABLED_LWM2M_FIRMWARE_OBJECT, inst, res);
+	if (ret < 0 || ret >= sizeof(path)) {
+		memset(path, 0, sizeof(path));
+	}
+	return path;
+}
+
+static int write_resource_to_settings(int inst, int res, uint8_t *data, uint16_t data_len)
+{
+	char path[sizeof(LWM2M_FIRM_PREFIX "/65632/0/10")];
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:190]:
```suggestion
	char path[sizeof(LWM2M_FIRM_PREFIX "/65535/0/10")];
```
Code-wise it doesn't really matter but made me to stop for a second to think why the value is what it is.

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c (hunk 6 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_adv_firmware.c:249]: True

**Comment by jheiskan81** [comment]
@VeijoPesonen I fixed your comment. I will update docs when code review is stable.

**Comment by jheiskan81** [comment]
@VeijoPesonen I added documentation about a new overlay

**Review by VeijoPesonen — APPROVED** [review]

File: samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+CONFIG_LWM2M_CLIENT_UTILS_ADV_FIRMWARE_UPDATE_OBJ_SUPPORT=y
+CONFIG_LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT=y
+CONFIG_LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT=n
+CONFIG_FOTA_CLIENT_EVT_DONE_SHEDULE_DFU=n
+# Application version
+CONFIG_MCUBOOT_IMAGE_VERSION="1.0.0"
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf:7]:
```suggestion
# Application version
CONFIG_MCUBOOT_IMAGE_VERSION="1.0.0"
```

Drop these.
Nothing to do with this object.

File: samples/nrf9160/lwm2m_client/sample_description.rst [hunk 9]
```diff
@@ -523,6 +523,7 @@ The following files are available:
 * :file:`overlay-assist-cell.conf` - Enables cell-based location assistance.
 * :file:`overlay-assist-pgps.conf` - Enables P-GPS assistance in the sample.
 * :file:`overlay-lowpower.conf` - Disables certain features to bring the power consumption down.
+* :file:`overlay-adv-firmware.conf` - Enable multi instance firmware update object. Modem and App have own instances.
```
Comment (SeppoTakalo) [inline samples/nrf9160/lwm2m_client/sample_description.rst:526]:
```suggestion
* :file:`overlay-adv-firmware.conf` - Enable experimental Advanced Firmware Update object.
```

File: subsys/net/lib/fota_download/Kconfig [hunk 10]
```diff
@@ -20,6 +20,10 @@ config FOTA_SOCKET_RETRIES
 config FOTA_DOWNLOAD_PROGRESS_EVT
 	bool "Emit progress event upon receiving a download fragment"
 
+config FOTA_CLIENT_EVT_DONE_SHEDULE_DFU
+	bool "Trigger DFU shedule after download ready state"
```
Comment (SeppoTakalo) [inline subsys/net/lib/fota_download/Kconfig:24]:
```suggestion
config FOTA_CLIENT_AUTOSCHEDULE_UPDATE
	bool "Mark target images to be updated when downloaded"
```

File: samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf (hunk 8 above)
Comment (jheiskan81) [reply samples/nrf9160/lwm2m_client/overlay-adv-firmware.conf:7]: I move this to Prj.conf

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 11]
```diff
@@ -131,77 +419,134 @@ static int configure_full_modem_update(void)
...
-					 update_counter.current + 1);
-	if (ret) {
-		LOG_ERR("Failed to update the update counter: %d", ret);
-		goto cleanup;
+	ret = lte_lc_deinit();
+	LOG_ERR("lte_lc_deinit() -> %d", ret);
+	ret = nrf_modem_lib_shutdown();
+	LOG_ERR("nrf_modem_lib_shutdown() -> %d", ret);
+	ret = nrf_modem_lib_init(NORMAL_MODE);
+	LOG_ERR("nrf_modem_lib_init(NORMAL_MODE) -> %d", ret);
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:443]:
```suggestion
	LOG_DBG("lte_lc_deinit() -> %d", ret);
	ret = nrf_modem_lib_shutdown();
	LOG_DBG("nrf_modem_lib_shutdown() -> %d", ret);
	ret = nrf_modem_lib_init(NORMAL_MODE);
	LOG_DBG("nrf_modem_lib_init(NORMAL_MODE) -> %d", ret);
```

Comment (SeppoTakalo) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:443]: or just remove..

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 12]
```diff
@@ -69,16 +95,285 @@ static void on_modem_lib_init(int ret, void *ctx)
...
+	dfu_image_type = target_image_type_get(obj_inst_id);
+	ret = firmware_target_pre_init(dfu_image_type);
+	if (ret) {
+		return ret;
+	}
+
+	return dfu_target_reset();
+}
+
+static int apply_firmware_update_type(int dfu_image_type)
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:333]:
```suggestion
static int set_firmware_update_type(int dfu_image_type)
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 13]
```diff
@@ -337,74 +689,87 @@ static int firmware_block_received_cb(uint16_t obj_inst_id,
...
 static void fota_download_callback(const struct fota_download_evt *evt)
 {
+	int dfu_image_type;
+
 	switch (evt->id) {
 	/* These two cases return immediately */
 	case FOTA_DOWNLOAD_EVT_PROGRESS:
-		LOG_DBG("DL progress %d", evt->progress);
+		if (ongoing_obj_id != UNUSED_OBJ_ID) {
+			/* Fet instance image type */
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:698]:
```suggestion
			/* Fetch instance image type */
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 14]
```diff
@@ -532,81 +906,243 @@ void lwm2m_firmware_set_update_state_cb(lwm2m_firmware_get_update_state_cb_t cb)
 	update_state_cb = cb;
 }
 
+static void lwm2m_firmware_object_store_read(int instance_id)
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:909]:
```suggestion
static void lwm2m_firmware_load_from_settings(int instance_id)
```

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c [hunk 15]
```diff
@@ -532,81 +906,243 @@ void lwm2m_firmware_set_update_state_cb(lwm2m_firmware_get_update_state_cb_t cb)
...
+	/* Read Object Spesific data */
+	snprintk(path, sizeof(path), LWM2M_FIRM_PREFIX "/%d/%d", ENABLED_LWM2M_FIRMWARE_OBJECT,
+		 instance_id);
+	ret = settings_load_subtree(path);
+	if (ret) {
+		LOG_ERR("Failed to load settings, %d", ret);
+	}
+}
+
+static void lwm2m_firmware_object_cb(int instance_id)
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:923]:
```suggestion
static void lwm2m_firmware_register_write_callbacks(int instance_id)
```

**Review by SeppoTakalo — CHANGES_REQUESTED** [review]

File: samples/nrf9160/lwm2m_client/sample_description.rst (hunk 9 above)
Comment (umapraseeda) [reply samples/nrf9160/lwm2m_client/sample_description.rst:526]: Just adding one more minor change on top of Seppo's suggestion.
```suggestion
* :file:`overlay-adv-firmware.conf` - Enables experimental Advanced Firmware Update object.
```

**Comment by umapraseeda** [comment]
@SeppoTakalo  Does this require a changelog entry? If not, the label could be removed.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c (hunk 11 above)
Comment (jheiskan81) [reply subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_firmware.c:443]: I remove traces.

**Comment by jheiskan81** [comment]
Thanks @SeppoTakalo and @umapraseeda I have fixed your comments.

**Comment by jheiskan81** [comment]
Fixed LwM2M utils unit test and Enabled Fota download progress event for old Firmware update to ATv2: @simensrostad

**Comment by jheiskan81** [comment]
/usr/local/lib/python3.8/site-packages/cmake/data/bin/cmake -E echo
/tools/ncs/toolchains/v2.1.0/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/zephyr_pre0.elf section `event_subscribers_all' will not fit in region `FLASH'
/tools/ncs/toolchains/v2.1.0/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: region `FLASH' overflowed by 40 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Was the flash usage at limit already @simensrostad

**Comment by jheiskan81** [comment]
I change FOTA_DOWNLOAD_PROGRESS_EVT enabling to LWM2M_CLIENT_UTILS_FIRMWARE_UPDATE_OBJ_SUPPORT if FOTA_DOWNLOAD. Now application user does not need to take care about this depency. So thats Why removed change form Lwm2m client sample and Asseset trackker v2. @SeppoTakalo  and @simensrostad Please comment how you feel. I was feeling that if app forget define FOTA_DOWNLOAD_PROGRESS_EVT.

**Review by umapraseeda — APPROVED** [review]

**Comment by SeppoTakalo** [comment]
> @SeppoTakalo Does this require a changelog entry? If not, the label could be removed.

This is experimental object, so we would like to not mention it yet. There is no server support for it, so this is going in only for the reason that once there is server support, we can start testing it against a real target. And only when it works, we claim that we support it.

So I would like to leave out the changelog, because it causes more confusion. No public documentation exist yet about this object and the definition is still subject to change.

**Comment by rlubos** [comment]
@jheiskan81 Please check the twister failure, this is a blocker for the PR.

**Comment by jheiskan81** [comment]
Any idea @rlubos  for fix this:
/tools/ncs/toolchains/v2.1.0/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/zephyr_pre0.elf section `event_subscribers_all' will not fit in region `FLASH'
/tools/ncs/toolchains/v2.1.0/opt/zephyr-sdk/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.bfd: region `FLASH' overflowed by 40 bytes
collect2: error: ld returned 1 exit status
I have a tiny affect for this

**Comment by rlubos** [comment]
Well, I don't know what would be the best way to fix the problem. The `applications.asset_tracker_v2.carrier` configuration must've been on the edge of FLASH already, given that your changes have triggered the overflow. I suggest contacting the application owners to see what could be optimized to fit again (or in the worst case disable configuration).

From my side, I'm just saying we can't merge a PR with a failing twister, as the failure would be permanent, meaning absolute havoc in the repository and other PRs.

**Comment by simensrostad** [comment]
@jheiskan81 I'll see if I can trim some memory.

File: applications/asset_tracker_v2/src/cloud/lwm2m_integration/lwm2m_integration.c [hunk 16]
```diff
@@ -295,6 +303,7 @@ static int firmware_update_state_cb(uint8_t update_state)
 		return 0;
 	case STATE_UPDATING:
 		LOG_DBG("STATE_UPDATING, result: %d", update_result);
+		send_firmware_update_done_event();
```
[the remark this answers is not in the document]
Comment (jheiskan81) [reply applications/asset_tracker_v2/src/cloud/lwm2m_integration/lwm2m_integration.c:306]: so now send operation needed?

**Comment by jheiskan81** [comment]
Thanks @simensrostad and @balaji-nordic for helping me. I have fixed unit test and cherry picked fix @balaji-nordic  to carrier library build. I fix @simensrostad also your comments.

**Review by simensrostad — APPROVED** [review]

**Comment by jheiskan81** [comment]
squash I fix commit to original

**Comment by jheiskan81** [comment]
Are ready for merge?

**Review by SeppoTakalo — APPROVED** [review]

**Comment by jheiskan81** [comment]
Rebuild test test-fw-nrfconnect-zigbee/ and test-fw-nrfconnect-chip Pass Please merge.

**Review by rlubos — APPROVED** [review]

**Pull request merged**